[PowerPC] Lower ucmp using subtractions

AZero13 opened the pull request:
Source: Hacker's delight, page 21.

Using the carry, we can use contractions to use the ucmp.

AZero13 commented:
A GI-Sel version is in the works. But this is just for Sel-DAG

AZero13 commented:
@jroelofs What do you think?

AZero13 commented:
@amy-kwan Thoughts?

AZero13 commented:
Note: I do not have merge permissions, so you can please merge if you approve this.

arsenm commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -12521,6 +12525,33 @@ SDValue PPCTargetLowering::LowerSSUBO(SDValue Op, SelectionDAG &DAG) const {
+
+  // Generate B - A using SUBC to capture carry.
+  SDVTList VTs = DAG.getVTList(OpVT, MVT::i32);
+  SDValue SubC = DAG.getNode(PPCISD::SUBC, DL, VTs, B, A);
```
Are PPC SUBC/SUBE just the same as the normal sub-with-overflow, just with glue? Can this be a generic expansion with those instead, and transform those into the target glue nodes?

AZero13 replied:
> Are PPC SUBC/SUBE just the same as the normal sub-with-overflow, just with glue? Can this be a generic expansion with those instead, and transform those into the target glue nodes? 

NO. 

Sube on PowerPC has a unique characteristic. It's not a-b-carry, it's a - b + carry. 

Because of this, only PowerPC gets to have a lowering like this.

AZero13 committed:
[PowerPC] Lower ucmp using subtractions

Source: Hacker's delight, page 21.

Using the carry, we can use contractions to use the ucmp.

SiliconA-Z commented:
> This is producing some kind of miscompilation, we are getting a rustc crash on PPC as a result [rust-lang/rust#151807](https://github.com/rust-lang/rust/issues/151807).

Ok let's see what the problem is.

SiliconA-Z commented:
Think I found it. Okay, let me try again!

tgross35 commented:
It's possible that this just happened to expose a bug that already existed. Any ideas how to reduce the issue further since it happens in a rather large binary?
